src/main/java/org/spongepowered/api/resource/ResourceReloadListener.java

```diff
@@ -0,0 +1,194 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.resource;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.data.persistence.DataFormat;
+import org.spongepowered.api.data.persistence.DataView;
+import org.spongepowered.api.event.resource.RegisterResourceReloadListenerEvent;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * A low level async reload listener.
+ *
+ * <p>There are two stages when reloading resources. The first is done
+ * asynchronously, so keep that in mind when interacting with game objects. The
+ * second is done synchronously on the game thread. This stage will not execute
+ * until all listeners have completed their first stage.</p>
+ *
+ * <p>In most circumstances, you should prefer to create a more specific
+ * use-case instance using the factory instead of implementing this interface
+ * directly.</p>
+ *
+ * <p>Example usage:</p>
+ *
+ * <pre>
+ *     public class MyAsyncReloadListener implements ReloadListener {
+ *         public CompletableFuture&lt;Void&gt; onReload(ReloadStage stage, ResourceManager manager, Executor workExecutor, Executor gameExecutor) {
+ *             return CompletableFuture
+ *                 // create a future in the work stage
+ *                 .supplyAsync(() -> {
+ *                     // load things async in work executor
+ *                     return things;
+ *                 }, workExecutor)
+ *                 // wait for other listeners to complete the work stage.
+ *                 .thenCompose(stage::markComplete)
+ *                 // create a future in the game stage
+ *                 .thenAcceptAsync(thing -> {
+ *                      // apply things sync in game executor
+ *                 }, gameExecutor);
+ *         }
+ *     }
+ * </pre>
+ *
+ * @see RegisterResourceReloadListenerEvent
+ */
+public interface ResourceReloadListener {
```

Comment:
Reading the javadocs, should it not contain two methods, one for the async stage and one for the sync stage, then just handle calling both methods elsewhere as appropriate? It seems to me that the javadocs say how reloading works, only to instruct you to do it yourself using `thenAccept` etc.

I feel like we should just tell devlopers to use the factories and not to implement this themselves.